src/server/service/GameService.ts

```diff
@@ -0,0 +1,229 @@
+    }
+
+    public getFreeRooms(): Room[] {
+        return this.roomRepository.getAllRooms().filter(room => room.players.length < 2);
+    }
+
+    public addShipsToRoom(playerIndex: string, shipsData: AddShipsRequest): boolean {
+        const room = this.roomRepository.getRoomByIndex(shipsData.gameId);
+        if (!room || room.isStarted) throw new Error('Room not found or battle already started');
+
+        const ships: Ship[] = shipsData.ships.map(ship => ({
+            position: ship.position,
+            direction: ship.direction,
+            length: ship.length,
+            type: ship.type,
+            hits: 0
+        }));
+        room.ships.set(playerIndex, ships);
+
+        return room.ships.size >= 2;
+    }
+
+    public startGame(roomIndex: string): Room {
+        const room = this.roomRepository.getRoomByIndex(roomIndex);
+        if (!room || room.isStarted) throw new Error('Room not found or battle already started');
+
+        room.isStarted = true;
+        this.switchToNextPlayer(room);
+        return room;
+    }
+
+    private executeAttack(playerIndex: string, roomIndex: string, x: number, y: number): AttackResult[] {
+        const attackResults: AttackResult[] = [];
+
+        const room = this.roomRepository.getRoomByIndex(roomIndex);
+        if (!room || !room.isStarted) throw new Error('Invalid room or game not started');
+        if (room.nextPlayerIndex !== playerIndex) throw new Error('Invalid player turn');
+
+        const opponentIndex = room.players.find(player => player.index !== playerIndex)?.index;
+        const opponentShips = room.ships.get(opponentIndex!);
+        if (!opponentShips) throw new Error('Opponent ships not found');
+
+        const hitShip = opponentShips.find(ship => this.isHit(ship, x, y));
+        if (!hitShip) {
+            attackResults.push(new AttackResult(x, y, 'miss'));
+            this.switchToNextPlayer(room);
+            return attackResults;
+        } else {
+            hitShip.hits++;
+            if (hitShip.hits >= hitShip.length) {
+                room.ships.set(opponentIndex!, opponentShips.filter(ship => ship !== hitShip));
+                this.addKilledForEachCell(hitShip, attackResults);
+                this.addMissesAroundShip(hitShip, attackResults);
+                return attackResults;
+            }
+        }
+        attackResults.push(new AttackResult(x, y, 'shot'));
+        return attackResults;
+    }
+
+    public attack(playerIndex: string, attackData: AttackRequest): AttackResult[] {
+        return this.executeAttack(playerIndex, attackData.gameId, attackData.x, attackData.y);
+    }
+
+    private isHit(ship: Ship, x: number, y: number): boolean {
+        return (
+            (ship.direction && ship.position.x === x && y >= ship.position.y && y < ship.position.y + ship.length) ||
+            (!ship.direction && ship.position.y === y && x >= ship.position.x && x < ship.position.x + ship.length)
+        );
+    }
```

Comment:
This logic is incomplete. You check if it is hit, but you do not check if you have hit this position previously. This is why I deducted your points